File: Έγγραφα SEM4105/Λύσεις ασκήσεων/08. Όγδοο εργαστήριο/1/lsq.f
      PROGRAM LEASQU
      INTEGER N
      DOUBLE PRECISION A1, A0, X(1000), Y(1000), R2

      INTEGER METHOD

      CALL READAT("points.dat", N, X, Y)

      PRINT *, "Μέθοδος: "
      PRINT *, "1) y = a * x + b"
      PRINT *, "2) y = a * x^b"
      PRINT *, "3) y = a + b * exp(x)"
      PRINT *, "4) y = a + b * ln(x)"
      PRINT *, "Δώσε μέθοδο."

      READ (*,*) METHOD
      
      IF (METHOD .EQ. 1) THEN 
         CALL LSQLIN(N, X, Y, A1, A0, R2)
         PRINT *, "y = ", A1, " x + ", A0
      ENDIF

      IF (METHOD .EQ. 2) THEN 
         CALL LSQPOW(N, X, Y, A1, A0, R2)
         PRINT *, "y = ", A0, " * x^", A1
      ENDIF

      IF (METHOD .EQ. 3) THEN 
         CALL LSQEXP(N, X, Y, A1, A0, R2)
         PRINT *, "y = ", A0, " + ", A1, " * exp(x)"
      ENDIF

      IF (METHOD .EQ. 4) THEN 
         CALL LSQLOG(N, X, Y, A1, A0, R2)
         PRINT *, "y = ", A0, " + ", A1, " * ln(x)"
      ENDIF
         
      PRINT *, "r^2 = ", R2

      END

C     Read data file
      SUBROUTINE READAT(FNAME, N, X, Y)
      CHARACTER(*) FNAME
      INTEGER N
      DOUBLE PRECISION X(*), Y(*)

      INTEGER UNITNO, I
      PARAMETER (UNITNO = 54)

      OPEN (UNIT = UNITNO, FILE = FNAME, STATUS = 'OLD')
      READ(UNITNO, *) N

      DO I=1,N
        READ (UNITNO, *) X(I), Y(I)
      ENDDO

      CLOSE (UNITNO)
      END


C     Y = A1 * X + A0
      SUBROUTINE LSQLIN(N, X, Y, A1, A0, R2)
      INTEGER N
      DOUBLE PRECISION X(*), Y(*), A1, A0, R2

      DOUBLE PRECISION SX, SY, SX2, SY2, SXY, D
      INTEGER I

      SX  = 0.0D0
      SY  = 0.0D0
      SX2 = 0.0D0
      SY2 = 0.0D0
      SXY = 0.0D0

      DO 10 I = 1, N
        SX  = SX  + X(I)
        SY  = SY  + Y(I)
        SX2 = SX2 + X(I) * X(I)
        SY2 = SY2 + Y(I) * Y(I)
        SXY = SXY + X(I) * Y(I)
 10   CONTINUE

      D = N * SX2 - SX * SX

      A1 = (N * SXY - SX * SY) / D 
      A0 = (SY - A1 * SX) / N

      R2 = (N * SXY - SX * SY)**2 / D / (N * SY2 - SY**2)
      END




C     Y = A0 X^A1 => LN(Y) = LN(A0) + A1 * LN(X)
      SUBROUTINE LSQPOW(N, X, Y, A1, A0, R2)
      INTEGER N
      DOUBLE PRECISION X(*), Y(*), A1, A0, R2

      DOUBLE PRECISION ZX(N), ZY(N), ZA0
      INTEGER I

      DO 10 I = 1, N
        IF ( (X(I) .LE. 0.D0) .OR. (Y(I) .LE. 0.D0) ) THEN
           PRINT *, "Data not appropriate"
           RETURN
        ENDIF

        ZX(I) = LOG(X(I))
        ZY(I) = LOG(Y(I))
 10   CONTINUE
      
      CALL LSQLIN(N, ZX, ZY, A1, ZA0, R2)

      A0 = EXP(ZA0)

      END


C     Y = A0 + A1 * EXP(X)
      SUBROUTINE LSQEXP(N, X, Y, A1, A0, R2)
      INTEGER N
      DOUBLE PRECISION X(*), Y(*), A1, A0, R2

      DOUBLE PRECISION ZX(N)
      INTEGER I

      DO 10 I = 1, N
        ZX(I) = EXP(X(I))
 10   CONTINUE
      
      CALL LSQLIN(N, ZX, Y, A1, A0, R2)

      END


C     Y = A0 + A1 * LOG(X)
      SUBROUTINE LSQLOG(N, X, Y, A1, A0, R2)
      INTEGER N
      DOUBLE PRECISION X(*), Y(*), A1, A0, R2

      DOUBLE PRECISION ZX(N)
      INTEGER I

      DO 10 I = 1, N
        IF ( (X(I) .LE. 0.0D0) ) THEN
           PRINT *, "Data not appropriate"
           RETURN
        ENDIF

        ZX(I) = LOG(X(I))
 10   CONTINUE
      
      CALL LSQLIN(N, ZX, Y, A1, A0, R2)
      
      END

File: Έγγραφα SEM4105/Λύσεις ασκήσεων/08. Όγδοο εργαστήριο/1/lsq.cpp
#include <vector>
#include <iostream>
#include <numeric>
#include <fstream>
#include <functional>
#include <cmath>
#include <algorithm>
#include <string>


using vec = std::vector<double>;

double mylog(double x)
{
    return std::log(x);
}

double myexp(double x)
{
    return std::exp(x);
}

bool allpositive(vec const & x)
{
    auto gt0 = [] (auto & x) -> bool { return x > 0; };
	
    return std::all_of(x.cbegin(), x.cend(), gt0);
}


//     Read data file
void read_data(std::string const & fname, vec & x, vec & y)
{
    std::ifstream in{fname};

    int n;
    in >> n;
    
    x.reserve(n);
    y.reserve(n);

    for (int i{0}; i < n; ++i) {
	double a, b;
	in >> a >> b;
	x.push_back(a);
	y.push_back(b);
    }
}


//     Y = A1 * X + a0
void lsqlin(vec const & x, vec const & y,
	    double & a1, double & a0, double & r2)
{
    auto const n = x.size();

    auto const sx = std::accumulate(x.cbegin(), x.cend(), 0.0);
    auto const sy = std::accumulate(y.cbegin(), y.cend(), 0.0);
    auto const sx2 = std::inner_product(x.cbegin(), x.cend(), x.cbegin(), 0.0);
    auto const sy2 = std::inner_product(y.cbegin(), y.cend(), y.cbegin(), 0.0);
    auto const sxy = std::inner_product(x.cbegin(), x.cend(), y.cbegin(), 0.0);

    auto const d = n * sx2 - sx * sx;

    a1 = (n * sxy - sx * sy) / d; 
    a0 = (sy - a1 * sx) / n;

    r2 = std::pow(n * sxy - sx * sy,2) / d / (n * sy2 - sy*sy);
}


//     Y = a0 X^A1 => LN(Y) = LN(a0) + A1 * LN(X)
void lsqpow(vec const & x, vec const & y,
	    double & a1, double & a0, double & r2)
{
    if (!allpositive(x) || !allpositive(y)) {
	std::cerr << "Data not appropriate\n";
	return; 
    }
    
    auto const n = x.size();
    vec zx(n);
    vec zy(n);
    double za0;

    std::transform(x.cbegin(), x.cend(), zx.begin(), mylog);
    std::transform(y.cbegin(), y.cend(), zy.begin(), mylog);    

    lsqlin(zx, zy, a1, za0, r2);
  
    a0 = std::exp(za0);
}


//     Y = a0 + a1 * LOG(X)
void lsqlog(vec const & x, vec const & y,
	    double & a1, double & a0, double & r2)
{
    if (!allpositive(x)) {
	std::cerr << "Data not appropriate\n";
	return; 
    }

    auto const n = x.size();
    vec zx(n);

    std::transform(x.cbegin(), x.cend(), zx.begin(), mylog);

    lsqlin(zx, y, a1, a0, r2);
  
}


//     Y = a0 + a1 * EXP(X)
void lsqexp(vec const & x, vec const & y,
	    double & a1, double & a0, double & r2)
{
    auto const n = x.size();
    vec zx(n);
        
    std::transform(x.cbegin(), x.cend(), zx.begin(), myexp);
    
    lsqlin(zx, y, a1, a0, r2);
}


int main()
{
    vec x, y;
    read_data("points.dat", x, y);

    std::cout <<  u8"Μέθοδος:\n"
	      << "1) y = a * x + b\n"
	      << "2) y = a * x^b\n"
	      << "3) y = a + b * exp(x)\n"
	      << "4) y = a + b * ln(x)\n"
	      << u8"Δώσε μέθοδο.\n";

    int method;
    std::cin >> method;

    double a1, a0, r2;

    switch (method) {
    case 1:
	lsqlin(x, y, a1, a0, r2);
	std::cout << "y = " << a1 << " x + " << a0 << '\n';
	break;
    case 2:
	lsqpow(x, y, a1, a0, r2);
	std::cout << "y = " << a0 << " * x^" << a1 << '\n';
	break;
    case 3:
	lsqexp(x, y, a1, a0, r2);
	std::cout << "y = " << a0 << " + " << a1 <<  " * exp(x)" <<'\n';
	break;
    case 4:
	lsqlog(x, y, a1, a0, r2);
	std::cout << "y = " << a0 << " + " << a1 << " * ln(x)" << '\n';
	break;
    }

    std::cout << "r^2 = " << r2 << '\n';
}


